File: src/dotnet/Ansight/AnsightOptionsBuilderExtensions.cs
namespace Ansight;

using Ansight.Tools.Database;
using Ansight.Tools.FileSystem;
using Ansight.Tools.Preferences;
using Ansight.Tools.Reflection;
using Ansight.Tools.SecureStorage;
using Ansight.Tools.VisualTree;
using System.Reflection;

/// <summary>
/// Provides all-in-one Ansight extensions for <see cref="Options.OptionsBuilder" />.
/// </summary>
public static class AnsightOptionsBuilderExtensions
{
    private const ushort DefaultSampleFrequencyMilliseconds = 400;
    private const ushort DefaultRetentionPeriodSeconds = 120;
    private const ushort DefaultSessionJpegCaptureIntervalMilliseconds = 2000;
    private const int DefaultSessionJpegCaptureQuality = 60;
    private const int DefaultSessionJpegCaptureMaxWidth = 480;

    /// <summary>
    /// Applies the default Ansight runtime configuration, registers the non-MAUI remote tools, and enables all tool scopes.
    /// </summary>
    /// <param name="builder">The Ansight options builder.</param>
    /// <returns>The current options builder.</returns>
    public static Options.OptionsBuilder WithAnsightSdk(this Options.OptionsBuilder builder)
    {
        ArgumentNullException.ThrowIfNull(builder);

        return builder.WithAnsightSdk(static _ => { });
    }

    /// <summary>
    /// Applies the default Ansight runtime configuration, enables all tool scopes, runs a customization callback,
    /// and registers the non-MAUI remote tools.
    /// </summary>
    /// <param name="builder">The Ansight options builder.</param>
    /// <param name="configure">
    /// A callback that customizes the existing Ansight options builder after runtime defaults and default tool access,
    /// but before default remote tool registration.
    /// Registering a tool suite inside this callback replaces the default all-in-one registration for that suite.
    /// </param>
    /// <returns>The current options builder.</returns>
    public static Options.OptionsBuilder WithAnsightSdk(
        this Options.OptionsBuilder builder,
        Action<Options.OptionsBuilder> configure)
    {
        ArgumentNullException.ThrowIfNull(builder);
        ArgumentNullException.ThrowIfNull(configure);

        builder = builder
            .WithAnsightDefaults()
            .WithAllToolAccess();
        configure(builder);
        return builder.WithAnsightRemoteTools();
    }

    /// <summary>
    /// Applies the default runtime configuration without registering remote tools or enabling tool access.
    /// </summary>
    /// <param name="builder">The Ansight options builder.</param>
    /// <returns>The current options builder.</returns>
    public static Options.OptionsBuilder WithAnsightDefaults(this Options.OptionsBuilder builder)
    {
        ArgumentNullException.ThrowIfNull(builder);

        builder = builder
            .WithFramesPerSecond()
            .WithSampleFrequencyMilliseconds(DefaultSampleFrequencyMilliseconds)
            .WithRetentionPeriodSeconds(DefaultRetentionPeriodSeconds)
            .WithSessionJpegCapture(
                DefaultSessionJpegCaptureIntervalMilliseconds,
                DefaultSessionJpegCaptureQuality,
                DefaultSessionJpegCaptureMaxWidth)
            .WithHostAutoProbe();

        var bundledHostConnectionAssembly = ResolveDefaultBundledHostConnectionAssembly();
        if (bundledHostConnectionAssembly is not null)
        {
            builder = builder.WithBundledHostConnection(bundledHostConnectionAssembly);
        }

#if IOS || MACCATALYST
        builder = builder.WithPlatformPairing();
#endif

        return builder;
    }

    /// <summary>
    /// Registers all non-MAUI remote tool suites, skipping any suite that already has a tool registered on the builder.
    /// </summary>
    /// <param name="builder">The Ansight options builder.</param>
    /// <returns>The current options builder.</returns>
    public static Options.OptionsBuilder WithAnsightRemoteTools(this Options.OptionsBuilder builder)
    {
        ArgumentNullException.ThrowIfNull(builder);

        if (!ContainsAnyTool(builder, visualTreeSuiteToolIds))
        {
            builder = builder.WithVisualTreeTools();
        }

        if (!ContainsAnyTool(builder, databaseSuiteToolIds))
        {
            builder = builder.WithDatabaseTools();
        }

        if (!ContainsAnyTool(builder, fileSystemSuiteToolIds))
        {
            builder = builder.WithFileSystemTools();
        }

        if (!ContainsAnyTool(builder, preferencesSuiteToolIds))
        {
            builder = builder.WithPreferencesTools();
        }

        if (!ContainsAnyTool(builder, reflectionSuiteToolIds))
        {
            builder = builder.WithReflectionTools();
        }

        if (!ContainsAnyTool(builder, secureStorageSuiteToolIds))
        {
            builder = builder.WithSecureStorageTools();
        }

        return builder;
    }

    private static bool ContainsAnyTool(Options.OptionsBuilder builder, IEnumerable<string> toolIds)
    {
        foreach (var toolId in toolIds)
        {
            if (builder.ContainsTool(toolId))
            {
                return true;
            }
        }

        return false;
    }

    private static Assembly? ResolveDefaultBundledHostConnectionAssembly()
        => Assembly.GetEntryAssembly();

    private static readonly string[] visualTreeSuiteToolIds =
    [
        VisualTreeToolIds.GetVisualTree,
        VisualTreeToolIds.GetScreenshot,
        VisualTreeToolIds.InspectNode,
        VisualTreeToolIds.ShowOverlay,
        VisualTreeToolIds.GetOverlay,
        VisualTreeToolIds.QueryOverlays,
        VisualTreeToolIds.UpdateOverlay,
        VisualTreeToolIds.RemoveOverlay,
        VisualTreeToolIds.ClearOverlays
    ];

    private static readonly string[] databaseSuiteToolIds =
    [
        DatabaseToolIds.ListDatabases,
        DatabaseToolIds.DescribeSchema,
        DatabaseToolIds.Query
    ];

    private static readonly string[] fileSystemSuiteToolIds =
    [
        FileSystemToolIds.ListDirectory,
        FileSystemToolIds.ReadFile,
        FileSystemToolIds.GetFileChecksum,
        FileSystemToolIds.DownloadFile,
        FileSystemToolIds.BeginBinaryDownload,
        FileSystemToolIds.PushFile,
        FileSystemToolIds.CopyFile,
        FileSystemToolIds.MoveFile,
        FileSystemToolIds.DeleteFile
    ];

    private static readonly string[] preferencesSuiteToolIds =
    [
        PreferencesToolIds.ListKeys,
        PreferencesToolIds.GetValue,
        PreferencesToolIds.SetValue,
        PreferencesToolIds.RemoveKey
    ];

    private static readonly string[] reflectionSuiteToolIds =
    [
        ReflectionToolIds.ListRoots,
        ReflectionToolIds.InspectObject,
        ReflectionToolIds.DescribeType,
        ReflectionToolIds.SetMemberValue,
        ReflectionToolIds.InvokeMethod
    ];

    private static readonly string[] secureStorageSuiteToolIds =
    [
        SecureStorageToolIds.GetValue,
        SecureStorageToolIds.SetValue,
        SecureStorageToolIds.RemoveKey
    ];
}

File: src/android/ansight/src/main/kotlin/ai/ansight/AnsightOptionsBuilderExtensions.kt
package ai.ansight

import ai.ansight.runtime.AnsightOptionsBuilder
import ai.ansight.tools.database.DatabaseToolIds
import ai.ansight.tools.database.withDatabaseTools
import ai.ansight.tools.filedescriptordiagnostics.FileDescriptorDiagnosticsToolIds
import ai.ansight.tools.filedescriptordiagnostics.withFileDescriptorDiagnosticsTools
import ai.ansight.tools.filesystem.FileSystemToolIds
import ai.ansight.tools.filesystem.withFileSystemTools
import ai.ansight.tools.preferences.PreferencesToolIds
import ai.ansight.tools.preferences.withPreferencesTools
import ai.ansight.tools.reflection.ReflectionToolIds
import ai.ansight.tools.reflection.withReflectionTools
import ai.ansight.tools.securestorage.SecureStorageToolIds
import ai.ansight.tools.securestorage.withSecureStorageTools
import ai.ansight.tools.visualtree.VisualTreeToolIds
import ai.ansight.tools.visualtree.withVisualTreeTools

fun AnsightOptionsBuilder.withAnsightSdk(): AnsightOptionsBuilder {
    return withAnsightSdk {}
}

fun AnsightOptionsBuilder.withAnsightSdk(
    configure: AnsightOptionsBuilder.() -> Unit,
): AnsightOptionsBuilder {
    withAnsightDefaults()
    withAllToolAccess()
    configure()
    return withAnsightRemoteTools()
}

fun AnsightOptionsBuilder.withAnsightDefaults(): AnsightOptionsBuilder {
    return withFramesPerSecond()
        .withSampleFrequencyMilliseconds(400)
        .withRetentionPeriodSeconds(120)
        .withSessionJpegCapture(
            intervalMilliseconds = 2_000,
            quality = 60,
            maxWidth = 480,
        )
        .withTouchCapture()
        .withHostAutoProbe()
}

fun AnsightOptionsBuilder.withAnsightRemoteTools(): AnsightOptionsBuilder {
    if (!containsAnyTool(visualTreeSuiteToolIds)) {
        withVisualTreeTools()
    }
    if (!containsAnyTool(databaseSuiteToolIds)) {
        withDatabaseTools()
    }
    if (!containsAnyTool(fileSystemSuiteToolIds)) {
        withFileSystemTools()
    }
    if (!containsAnyTool(fileDescriptorDiagnosticsSuiteToolIds)) {
        withFileDescriptorDiagnosticsTools()
    }
    if (!containsAnyTool(preferencesSuiteToolIds)) {
        withPreferencesTools()
    }
    if (!containsAnyTool(reflectionSuiteToolIds)) {
        withReflectionTools()
    }
    if (!containsAnyTool(secureStorageSuiteToolIds)) {
        withSecureStorageTools()
    }

    return this
}

private fun AnsightOptionsBuilder.containsAnyTool(toolIds: Iterable<String>): Boolean {
    return toolIds.any { containsTool(it) }
}

private val visualTreeSuiteToolIds = listOf(
    VisualTreeToolIds.GetVisualTree,
    VisualTreeToolIds.GetScreenshot,
    VisualTreeToolIds.InspectNode,
    VisualTreeToolIds.ShowOverlay,
    VisualTreeToolIds.GetOverlay,
    VisualTreeToolIds.QueryOverlays,
    VisualTreeToolIds.UpdateOverlay,
    VisualTreeToolIds.RemoveOverlay,
    VisualTreeToolIds.ClearOverlays,
)

private val databaseSuiteToolIds = listOf(
    DatabaseToolIds.ListDatabases,
    DatabaseToolIds.DescribeSchema,
    DatabaseToolIds.Query,
)

private val fileSystemSuiteToolIds = listOf(
    FileSystemToolIds.ListDirectory,
    FileSystemToolIds.ReadFile,
    FileSystemToolIds.GetFileChecksum,
    FileSystemToolIds.DownloadFile,
    FileSystemToolIds.BeginBinaryDownload,
    FileSystemToolIds.PushFile,
    FileSystemToolIds.CopyFile,
    FileSystemToolIds.MoveFile,
    FileSystemToolIds.DeleteFile,
)

private val fileDescriptorDiagnosticsSuiteToolIds = listOf(
    FileDescriptorDiagnosticsToolIds.ListOpen,
    FileDescriptorDiagnosticsToolIds.CountOpen,
    FileDescriptorDiagnosticsToolIds.Inspect,
    FileDescriptorDiagnosticsToolIds.GetUsage,
)

private val preferencesSuiteToolIds = listOf(
    PreferencesToolIds.ListKeys,
    PreferencesToolIds.GetValue,
    PreferencesToolIds.SetValue,
    PreferencesToolIds.RemoveKey,
)

private val reflectionSuiteToolIds = listOf(
    ReflectionToolIds.ListRoots,
    ReflectionToolIds.InspectObject,
    ReflectionToolIds.DescribeType,
    ReflectionToolIds.SetMemberValue,
    ReflectionToolIds.InvokeMethod,
)

private val secureStorageSuiteToolIds = listOf(
    SecureStorageToolIds.GetValue,
    SecureStorageToolIds.SetValue,
    SecureStorageToolIds.RemoveKey,
)
